### PR TITLE
Migrating package manager from pip to Anaconda

### LINKED ISSUE #84: Migrate package requirements to Anaconda
Right now our `requirements.txt` is broken, it doesn't work on a fresh clone of the repo. We should be able to:

1. Clone repo
2. Install packages from requirements
3. Run all of our scripts, notebooks, etc. without issue

We'll switch from pip's `requirements.txt` approach to using the Anaconda package manager

### PR DESCRIPTION
**Pull request by willfenton**
I've removed the `requirements.txt` file and replaced it with `environment.yml`, which is Anaconda's exported environment format. 

I'd appreciate other people testing it out and running their code with it to make sure it works. I've only tested it on Linux, so I'm not sure if it will work on other operating systems. See `README.md` for install instructions.

### CODE REVIEW

**Comment by ExiaSR** [comment]
@willfenton 
tldr, try using <https://github.com/ContinuumIO/anaconda-issues/issues/9480#issuecomment-444117929> to create the yml file

I got
```
(base) ▲ (michael-lin1.local) CANOSP2020 [will-conda-migration] conda env create -f environment.yml      
Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound: 
  - sqlite==3.31.1=h7b6447c_0
  - expat==2.2.6=he6710b0_0
  - ujson==1.35=py37h14c3975_0
  - c-ares==1.15.0=h7b6447c_1001
  - dbus==1.13.12=h746ee38_0
  - gstreamer==1.14.0=hb453b48_1
  - zeromq==4.3.1=he6710b0_3
  - intel-openmp==2019.4=243
  - scipy==1.1.0=py37hfa4b5c9_1
  - libpng==1.6.37=hbc83047_0
  - wrapt==1.10.11=py37h14c3975_2
  - gst-plugins-base==1.14.0=hbbd80ab_1
  - pyrsistent==0.15.7=py37h7b6447c_0
  - regex==2020.1.8=py37h7b6447c_0
  - zlib==1.2.11=h7b6447c_3
  - gmp==6.1.2=h6c8ec71_1
  - markupsafe==1.1.1=py37h7b6447c_0
  - python==3.7.6=h0371630_2
  - grpcio==1.27.2=py37hf8bcb03_0
  - yaml==0.1.7=had09818_2
  - scikit-learn==0.20.1=py37h4989274_0
  - pcre==8.43=he6710b0_0
  - cryptography==2.8=py37h1ba5d50_0
  - pyqt==5.9.2=py37h05f1152_2
  - fontconfig==2.13.0=h9420a91_0
  - libgcc-ng==9.1.0=hdf63c60_0
  - mistune==0.8.4=py37h7b6447c_0
  - mkl-service==1.1.2=py37h90e4bf4_5
  - libxml2==2.9.9=hea5a465_1
  - tensorboard==1.13.1=py37hf484d3e_0
  - libstdcxx-ng==9.1.0=hdf63c60_0
  - libsodium==1.0.16=h1bed415_0
  - icu==58.2=h9c2bf20_1
  - libuuid==1.0.3=h1bed415_2
  - pyzmq==18.1.1=py37he6710b0_0
  - pyyaml==5.3=py37h7b6447c_0
  - qt==5.9.7=h5867ecd_1
  - jpeg==9b=h024ee3a_2
  - freetype==2.9.1=h8a8886c_1
  - cytoolz==0.9.0.1=py37h14c3975_1
  - mkl_fft==1.0.6=py37h7dd41cf_0
  - xz==5.2.4=h14c3975_4
  - tornado==6.0.3=py37h7b6447c_3
  - murmurhash==1.0.2=py37he6710b0_0
  - sip==4.19.8=py37hf484d3e_0
  - cffi==1.14.0=py37h2e261b9_0
  - libgfortran-ng==7.3.0=hdf63c60_0
  - pandas==1.0.1=py37h0573a6f_0
  - ncurses==6.2=he6710b0_0
  - h5py==2.10.0=py37h7918eee_0
  - mkl_random==1.0.1=py37h4414c95_1
  - openssl==1.1.1d=h7b6447c_4
  - kiwisolver==1.1.0=py37he6710b0_0
  - ld_impl_linux-64==2.33.1=h53a641e_7
  - libprotobuf==3.11.4=hd408876_0
  - libedit==3.1.20181209=hc058e9b_0
  - tensorflow-base==1.13.1=mkl_py37h7ce6ba3_0
  - libxcb==1.13=h1bed415_1
  - cymem==2.0.2=py37hfd86e86_0
  - tbb==2020.0=hfd86e86_0
  - readline==7.0=h7b6447c_5
  - tk==8.6.8=hbc83047_0
  - msgpack-python==0.6.1=py37hfd86e86_1
  - glib==2.63.1=h5a9c865_0
  - hdf5==1.10.4=hb1b8bf9_0
  - matplotlib-base==3.1.3=py37hef1b27d_0
  - tensorflow==1.13.1=mkl_py37h54b294f_0
  - protobuf==3.11.4=py37he6710b0_0
  - libffi==3.2.1=hd88cf55_4
  - tbb4py==2020.0=py37hfd86e86_0
```

**Comment by ExiaSR** [comment]
Another idea,

Use github actions to run conda install on different environments make sure every commit actually work on fresh clone.

https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idruns-on

**Comment by mlopatka** [comment]
I am also unable to conda install with the current configuration.

`(base) mlopatka-fmjhd3:CANOSP2020 mlopatka$ conda env create --file=environment.yml 
Collecting package metadata: done
Solving environment: failed

ResolvePackageNotFound: 
  - tensorflow==1.13.1=mkl_py37h54b294f_0
  - protobuf==3.11.4=py37he6710b0_0
  - xz==5.2.4=h14c3975_4
  - msgpack-python==0.6.1=py37hfd86e86_1
  - expat==2.2.6=he6710b0_0
  - libuuid==1.0.3=h1bed415_2
  - h5py==2.10.0=py37h7918eee_0
...
`

File: environment.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  - spacy
+  - tensorflow
+  - pip:
+    - cld2-cffi
+    - crowdtruth
+    - nmslib
+    - scikit-multilearn
+    - textpipe
+    - yake
+prefix: /home/wfenton/anaconda3/envs/canosp
```
Comment (ExiaSR) [inline environment.yml]: Hardcoded `prefix` is probably a bad idea

**Review by michaellzc — CHANGES_REQUESTED** [review]

**Comment by birdsarah** [comment]
Remove prefix as per @ExiaSR and add back in windows and ping me if it's still failing.

**Comment by willfenton** [comment]
Removed the prefix line, same error `[WinError 5] Access is denied`. It seems like solution is to use the `--user` flag when installing pip packages, but after doing some research [it doesn't seem like you can set pip flags](https://github.com/conda/conda/issues/6805) in the `environment.yml` file. 

One option would be to create a separate `requirements.txt` with our pip requirements and install like so, but since it's currently working on Ubuntu and MacOS, that's not ideal.

```
conda env create -f environment.yml
conda activate canosp
pip install --user -r requirements.txt
```


Windows log: [(Link)](https://github.com/mlopatka/CANOSP2020/runs/482991074)
```
Pip subprocess error:

  WARNING: Legacy build of wheel for 'textpipe' created no files.

  Command arguments: 'C:\Miniconda\envs\canosp\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\pip-install-to18v7aq\\textpipe\\setup.py'"'"'; __file__='"'"'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\pip-install-to18v7aq\\textpipe\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\RUNNER~1\AppData\Local\Temp\pip-wheel-0uijtl58'

  Command output: [use --verbose to show]

ERROR: Could not install packages due to an EnvironmentError: [WinError 5] Access is denied: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\pip-uninstall-9zds_eb9\\tqdm.exe'

Consider using the `--user` option or check the permissions.





CondaEnvException: Pip failed

##[error]Process completed with exit code 1.

```

@birdsarah

**Comment by birdsarah** [comment]
I don't know but I don't *think* the user thing is the issue. I *think* that issue is coming from the inability to uninstall tqdm which is necessitated by the fact that textpipe can't be built as a wheel for some reason.

add `-vvv` to the conda env create line to see if we can get any more info.

After that, lets try leaving everything the same except remove textpipe from the environment.yaml and add `pip install textpipe` (after conda activate step, to see where that lands us.)

**Comment by birdsarah** [comment]
I should add that i don't know why it wants to be built as a wheel and/or how to stop that.

**Comment by birdsarah** [comment]
My understanding from the linked conda issue is may be able to try `--install-option` flag inline in environment.yaml

https://github.com/conda/conda/issues/6805#issuecomment-361782555

In environment.yaml
```yaml
 .....
  - setuptools
  - spacy
  - tensorflow
  - pip:
    - cld2-cffi
    - crowdtruth
    - nmslib
    - scikit-multilearn
    - textpipe --install-option=""
    - yake
```

**Comment by birdsarah** [comment]
separate problem. this environment.yaml is currently installing spacy in conda and pip as far as i can tell. that will definitely cause you an issue.

try this pr without the other modifications and just using conda-forge channel which has a much more up to date version of spacy instead of defaults. (https://anaconda.org/search?q=spacy)

also check the logs in detail to make sure that other packages are not being dual installed by your pip requirements.

**Comment by birdsarah** [comment]
try adding tqdm into the envrionment.yaml as a conda dependency

**Comment by birdsarah** [comment]
i'll awaken my windows bo

**Comment by birdsarah** [comment]
OK. There's lots going on here. Let's split this up into constituent parts instead of me stabbing in the dark more

Lets get convinced that textpipe is pip installable on the github workflows infrastructure:
* Please open a separate PR that just adds a github workflow that pip installs the dependencies. 
* OR point me to where this already exists.

**Comment by birdsarah** [comment]
Other misc observations:
* conda install cld2-cffi
* scikit-multilearn - does not appear on quick glance to be up-to-date and well supported. What are the parameters for package selection and verification?
* pathlib - old and unecessary, why are you not using python standard library?
* setuptools - i am not a super whiz at python packaging, so if there's a good reason for inclusion then great, but it's not a library i usually include in my conda environment
* yake - pip installable version is listed as unmaintained - update installation (https://pypi.org/project/yake/, https://github.com/LIAAD/yake)

**Comment by willfenton** [comment]
* Good call with `cld2-cffi`, also with `pathlib`. I actually didn't realize that it was included in the standard library
* Requirement for including a package was that we used it somewhere already in our code. All of our existing code should run with the new Conda environment. We're using `scikit-multilearn` in one of our Jupyter Notebooks.
* We're using [Char2Vec](https://github.com/Lettria/Char2Vec), and so I included anything from its `requirements.txt` in our `environment.yml`, such as `setuptools`

As for pip installing textpipe on github workflows, I'll test that.

Thank you so much for helping out, @birdsarah!

**Comment by willfenton** [comment]
Don't let the :heavy_check_mark:  fool you, still can't get textpipe to work. `Failed to build textpipe`

Although it failed to build on all operating systems this time. At least the environment sans textpipe works on all of them, so progress?

Will investigate further, maybe not in the next couple days though since I have a ton of other work to do.

**Comment by ExiaSR** [comment]
@willfenton We got rid of textpipe from our dependency list.

Can you get a working conda environment setup with the following libraries

- nltk
- spacy
- genism
- sklearn

Anything else?

**Comment by willfenton** [comment]
This is now working on all operating systems, I think we're ready to merge.

**Comment by mlopatka** [comment]
nice work @willfenton 
Please squash commits before merging in.

**Review by birdsarah — APPROVED** [review]
Looks fine

**Review by mlopatka — APPROVED** [review]

**Review by michaellzc — APPROVED** [review]

File: testfile [hunk 2]
```diff
@@ -0,0 +1,2 @@
+testing file for github actions
```
Comment (ExiaSR) [inline testfile]: Remove this file

**Review by michaellzc — CHANGES_REQUESTED** [review]

File: testfile (hunk 2 above)
Comment (willfenton) [reply testfile]: Thanks, good catch

**Review by michaellzc — APPROVED** [review]

File: .github/workflows/conda.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+name: Setup Anaconda environment
+
+on:
+  pull_request:
+    branches:
```
Comment (ExiaSR) [inline .github/workflows/conda.yml]: One more thing, could you config this workflow only run when the `environment.yml` file changes?

Comment (willfenton) [reply .github/workflows/conda.yml]: I briefly looked into it earlier and I don't think you can, I'll look into it though
